discovery-azure-api/src/main/scala/akka/discovery/azureapi/AzureRbacAksServiceDiscovery.scala

```diff
@@ -0,0 +1,264 @@
+
+  private val accessTokenRequestContext: TokenRequestContext =
+    new TokenRequestContext()
+
+  /**
+   * INTERNAL API
+   *
+   * Finds relevant targets given a pod list. Note that this doesn't filter by name as it is the job of the selector
+   * to do that.
+   */
+  @InternalApi
+  private[azureapi] def targets(
+      podList: PodList,
+      portName: Option[String],
+      podNamespace: String,
+      podDomain: String,
+      rawIp: Boolean,
+      containerName: Option[String]): immutable.Seq[ResolvedTarget] =
+    for {
+      item <- podList.items
+      if item.metadata.flatMap(_.deletionTimestamp).isEmpty
+      itemSpec <- item.spec.toSeq
+      itemStatus <- item.status.toSeq
+      if itemStatus.phase.contains("Running")
+      if containerName.forall(name =>
+        itemStatus.containerStatuses match {
+          case Some(statuses) => statuses.filter(_.name == name).exists(!_.state.contains("waiting"))
+          case None           => false
+        })
+      ip <- itemStatus.podIP.toSeq
+      // Maybe port is an Option of a port, and will be None if no portName was requested
+      maybePort <- portName match {
+        case None =>
+          List(None)
+        case Some(name) =>
+          for {
+            container <- itemSpec.containers
+            ports <- container.ports.toSeq
+            port <- ports
+            if port.name.contains(name)
+          } yield Some(port.containerPort)
+      }
+    } yield {
+      val hostOrIp = if (rawIp) ip else s"${ip.replace('.', '-')}.$podNamespace.pod.$podDomain"
+      ResolvedTarget(
+        host = hostOrIp,
+        port = maybePort,
+        address = Some(InetAddress.getByName(ip))
+      )
+    }
+
+  final class KubernetesApiException(msg: String) extends RuntimeException(msg) with NoStackTrace
+
+  final case class KubernetesSetup(namespace: String, ctx: HttpsConnectionContext)
+}
+
+class AzureRbacAksServiceDiscovery(implicit system: ExtendedActorSystem) extends ServiceDiscovery {
+  private val http = Http()
+
+  private val settings = Settings(system)
+
+  private val log = Logging(system, classOf[AzureRbacAksServiceDiscovery])
+
+  log.debug("Settings {}", settings)
+
+  private def fetchToken: Future[AccessToken] =
+    azureDefaultCredential.getToken(accessTokenRequestContext.addScopes(settings.entraServerId)).toFuture.asScala
+
+  private val kubernetesSetup = {
+    implicit val blockingDispatcher: ExecutionContext = system.dispatchers.lookup(DefaultBlockingDispatcherId)
+    for {
+      namespace: String <- Future {
+        settings.podNamespace
+          .orElse(readConfigVarFromFilesystem(settings.podNamespacePath, "pod-namespace"))
+          .getOrElse("default")
+      }
+      httpsContext <- Future(clientHttpsConnectionContext())
+    } yield {
+      KubernetesSetup(namespace, httpsContext)
+    }
+  }
+
+  import system.dispatcher
+
+  private def accessToken(retries: Int = 3): Future[AccessToken] = {
```

Comment:
> I would assume that azureDefaultCredential.getToken is stable enough to use without having to retry. If we need to retry then I would like to see a back off strategy.

I'll find out.

Reply:
> I would assume that azureDefaultCredential.getToken is stable enough to use without having to retry. If we need to retry then I would like to see a back off strategy.

`getToken` is called automatically by Azure SDK client libraries but since we aren't using them to communicate with the aks api server we'd have to handle caching, token refreshing, and retries.

The library will just emit `CredentialUnavailableException` if it is unable to get the access token for any reason.